cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java

```diff
@@ -1016,6 +1027,55 @@ protected Future<Void> updateKafkaVersion(KafkaStatus kafkaStatus) {
+        return Future.succeededFuture();
+    }
+
+    /**
+     * This method checks if a migration is still ongoing on the Kafka side, through the KafkaMetadataStateManager instance.
+     * A ZooKeeper to KRaft migration can take some time and, on each reconcile, the operator checks its status by calling this method.
+     * Internally, the KafkaMetadataStateManager instance is leveraging the endpoint exposed by the Kafka Agent which provides
+     * the KRaft migration state through a corresponding metric.
+     *
+     * @return  Future which completes when the check on the migration is done
+     */
+    protected Future<Void> updateKafkaMetadataMigrationState() {
+        KafkaMetadataConfigurationState kafkaMetadataConfigState = this.kafkaMetadataStateManager.getMetadataConfigurationState();
+        // on each reconcile, would be useless to check migration status if it's not going on
+        if (kafkaMetadataConfigState.isMigration()) {
+            return ReconcilerUtils.clientSecrets(reconciliation, secretOperator)
+                    .compose(compositeFuture -> {
+                        // we should get the quorum controller leader using the Admin Client API describeMetadataQuorum() but
+                        // it's not supported by brokers during migration because they are still connected to ZooKeeper so ...
+                        // going through the controllers set to get metrics from one of them, because all expose the needed metrics
+                        boolean zkMigrationStateChecked = false;
+                        for (NodeRef controller : kafka.controllerNodes()) {
+                            int controllerId = controller.nodeId();
+                            String controllerPodName = kafka.nodePoolForNodeId(controllerId).nodeRef(controllerId).podName();
+                            try {
+                                LOGGER.infoCr(reconciliation, "Checking ZooKeeper migration state on controller {}", controllerPodName);
+                                this.kafkaMetadataStateManager.checkMigrationInProgress(
+                                        reconciliation,
+                                        compositeFuture.resultAt(0),
+                                        compositeFuture.resultAt(1),
+                                        controllerPodName
+                                );
+                                zkMigrationStateChecked = true;
+                                break;
+                            } catch (RuntimeException e) {
+                                LOGGER.warnCr(reconciliation, "Error on checking ZooKeeper migration state on controller {}", controllerPodName);
+                            }
+                        }
+                        return zkMigrationStateChecked ? Future.succeededFuture() : Future.failedFuture(new Throwable("Impossible to check ZooKeeper migration state"));
```

Comment:
I don't follow this logic. You loop through all brokers but jump out after checking the first one? You also absolutely don't care about what the state outcome is and call it a success just if one of the nodes doesn't fail with an exception?Why do you actually do this?

Reply:
Let me explain the logic behind this ...
What we need is providing a controller (pod)name to the `checkMigrationInProgress` which is going to query the Kafka agent on that pod about the migration state metric.
At the beginning, the idea was to use the quorum controller leader to ask that metric but you can see the comment before the loop ...

```
// we should get the quorum controller leader using the Admin Client API describeMetadataQuorum() but
// it's not supported by brokers during migration because they are still connected to ZooKeeper
```

Luckily the metric is exposed by all the controllers so picking just one is fine.
My first implementation was about picking a random one but after your review I changed to go through all of them until one of them is able to provide me the metric value.
If one of them is not reachable, the exception is caught and we move forward to the next one. As soon as one of them is able to reply so without exceptions, it means that the `checkMigrationInProgress` was able to get the migration state metric and we know at which stage the migration is, so state is checked and we can exit.
Of course, if no controllers were able to provide the metric, the method will exit with the failed future.

Reply:
I think it might be much more readable if you change the code to _do something_ => get some output from the node, check it, move the state manage, and break. This way it looks like it just doesn't do anything.

Reply:
I moved the check of migration state out of the state manager (to the Utils class). This loop now calls the Utils class and set the result to the state manager so that it can evaluate the state transitions. I guess now it's more understandable making clearer which "action" is running.